File: stacks-and-queues/valid_parentheses.cpp
// 20. Valid Parentheses
// https://leetcode.com/problems/valid-parentheses/description/
// Difficulty: Easy
// Strategy: stack.
#include <iostream>
#include <stack>
#include <unordered_map>

using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> stack;
        unordered_map<char, char> matching {{'(',')'}, {'{','}'}, {'[',']'}};

        for(char c : s) {
            if (matching.find(c) != matching.end()) {
                stack.push(c);
            } else {

                if (stack.empty()) {
                    return false;
                }

                char lastItem = stack.top();
                if (matching[lastItem] != c) {
                    return false;
                }

                stack.pop();
            }
            
        }
        
        return stack.empty();

    }
};

int main() {
    Solution sol;
    string s = "()[]{}";
    cout << (sol.isValid(s) ? "true" : "false")<< "\n";
    return 0;
}

File: stacks-and-queues/valid_parentheses.py
# 20. Valid parenthesis
# https://leetcode.com/problems/valid-parentheses/description/
# Difficulty: Easy
# Strategy: Stack.

class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        parenthesis = {"(": ")", "[": "]", "{": "}"}

        for p in s:
            if stack:
                top = stack[-1]
                if top in parenthesis and parenthesis[top] == p:
                    stack.pop()
                else: 
                    stack.append(p)
            else:
                stack.append(p)

        if stack:
            return False
        else:
            return True
        

def test():
    sol = Solution()
    s = "([)]"
    return sol.isValid(s)

print(test())
